app/models/changeset.rb

```diff
@@ -67,53 +66,54 @@ def error
   private
 
   def find_comparison
-    return NullComparison.new(nil) if empty?
-
-    # for branches that frequently change we make sure to always get the correct cache,
-    # others might get an outdated changeset if they are reviewed with different shas
-    if BRANCH_TAGS.include?(@reference)
-      @reference = Samson::Hooks.fire(:changeset_api_request, self, :branch).compact.first
-    end
+    return NullComparison.new if empty?
 
-    Rails.cache.fetch(cache_key) do
-      Samson::Hooks.fire(:changeset_api_request, self, :compare).compact.first
+    Rails.cache.fetch(compare_cache_key) do
+      Samson::Hooks.fire(:repo_compare, @project, @previous_commit, @reference).compact.first
     end
   rescue StandardError => e
-    Changeset::NullComparison.new(e.message)
+    Changeset::NullComparison.new(error: "Repository error: #{e.message.split("::").last}")
   end
 
-  def find_pull_requests
-    numbers = commits.map(&:pull_request_number).compact
-    numbers.map { |num| PullRequest.find(@repo, num) }.
-      compact.
-      concat(find_pull_requests_for_branch)
+  def merged_pull_requests
+    commits.map(&:pull_request_number).compact.map { |num| PullRequest.find(@repo, num) }.compact
   end
 
-  def cache_key
-    [self.class, @repo, @previous_commit, @reference].join('-')
+  # for branches that frequently change we make sure to always get the correct cache,
+  # others might get an outdated changeset if they are reviewed with different shas
+  # it then still does a http request, but it is much faster
+  def compare_cache_key
+    key =
+      if static?
+        @reference
+      else
+        @project.repo_commit_from_ref(@reference)
+      end
+
+    [self.class, @repo, @previous_commit, key].join('-')
   end
 
   # github only supports finding open PRs for branches (not commits or tags)
   # https://help.github.com/en/articles/searching-issues-and-pull-requests
-  def find_pull_requests_for_branch
-    return [] if not_pr_branch?
+  def open_pull_requests
+    return [] if static? || MAIN_BRANCHES.include?(@reference)
```

Comment:
is @reference guaranteed to be a branch here?

Reply:
what if it's a tag?

Reply:
`static?` checks that
... it's just optimization to avoid doing the request